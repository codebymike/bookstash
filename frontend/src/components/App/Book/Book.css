div#book_page{
  padding:10px;
}

div#book_page img.main{
  float:right;
  width:300px;
}

div#book_page dl{
  margin:20px 0 10px;
  width: 50%;
}

div#book_page dl dt{
  float:left;
  width: 50%;
  font-weight: bold;
  clear:left;
  margin-bottom:5px;
}

div#book_page dl dd{
  float:right;
  width: 50%;
  margin-bottom:5px;
}
