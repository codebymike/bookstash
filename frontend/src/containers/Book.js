import React from 'react'
import { connect } from 'react-redux'
import BookView from '../components/BookView'
import { getBook } from '../actions'

class Book extends React.Component {

  componentDidMount() {
    this.props.dispatch( getBook('0307347974') );
  }

  render() {
    return (
      <BookView book={this.props.book}/>
    )
  }
}

const mapStateToProps = state => ({
  book: state.book
});

export default connect(mapStateToProps)(Book);
