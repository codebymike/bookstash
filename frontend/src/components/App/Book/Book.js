import React from 'react';
import './Book.css';

class Book extends React.Component{

  state = {
    book: {}
  }

  componentDidMount = () => {
    // this.fetchBooks( this.props.url );
    this.fetchBook( this.props.match.params.book_id );
  }

  fetchBook = async (id) => {
    fetch( `https://b1w5pwo5bd.execute-api.eu-west-1.amazonaws.com/latest/books/${ id }` )
      .then(response => response.json())
      .then(book => this.setState({ book }))
      .catch(error => console.log(error));
  }

  render = () => {

    const { title, image, author, rating, ratings, length, first_published, isbn } = this.state.book;

    return (
      <div id="book_page">
        <h1>{ title }</h1>
        <img src={ image } alt={ title }/>
        <dl>
          <dt>Author</dt><dd>{ author }</dd>
          <dt>First Published</dt><dd>{ first_published }</dd>
          <dt>Rating</dt><dd>{ rating }</dd>
          <dt>Ratings</dt><dd>{ ratings }</dd>
          <dt>Length</dt><dd>{ length }</dd>
          <dt>ISBN</dt><dd>{ isbn }</dd>
        </dl>
      </div>
    )
  }


}

export default Book;
