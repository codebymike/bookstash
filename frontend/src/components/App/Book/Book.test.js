import React from 'react';
import { shallow } from 'enzyme';
import Book from './Book';

// global.fetch = require('jest-fetch-mock');
// Manual mocking
const fetchPromise = Promise.resolve({
  json: () => Promise.resolve(
    {
      "title": "Fahrenheit 451",
      "author": "Ray Bradbury",
      "image": "https://images.gr-assets.com/books/1351643740l/4381.jpg",
      "rating": "3.98",
      "ratings": 1266,
      "length": 175,
      "first_published": 1953,
      "isbn": "0307347974",
      "url": "https://www.goodreads.com/book/show/4381.Fahrenheit_451"
    }
  ),
})
global.fetch = () => fetchPromise

const sleep = (ms) => {
  return new Promise(resolve => setTimeout(resolve, ms));
}

describe('Book Component', () => {

  it('should render without throwing an error', () => {
    expect(shallow( <Book match={{params: {book_id: '0307347974'}}} /> ).find('div#book_page').exists()).toBe(true)
  })

  it('should update the state when mounted', async () => {

    const book = shallow(<Book match={{params: {book_id: '0307347974'}}} />)

    await book.instance().componentDidMount()
    await sleep(500) // wait for fetchBook to update the state

    expect(book.state('book')).toHaveProperty('title', 'Fahrenheit 451')

  })

})
