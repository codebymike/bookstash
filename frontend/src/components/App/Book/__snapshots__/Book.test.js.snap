// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book Component should match the snapshot 1`] = `
<div
  id="book_page"
>
  <img
    alt={undefined}
    className="main"
    src={undefined}
  />
  <h1 />
  <h2>
    by 
  </h2>
  <dl>
    <dt>
      First Published
    </dt>
    <dd />
    <dt>
      Rating
    </dt>
    <dd>
      /5
    </dd>
    <dt>
      Number of Ratings
    </dt>
    <dd />
    <dt>
      Length
    </dt>
    <dd>
       pages
    </dd>
    <dt>
      ISBN
    </dt>
    <dd />
  </dl>
</div>
`;
